File: backend/src/models/Attendance.js
const mongoose = require('mongoose');

const AttendanceRecordSchema = new mongoose.Schema({
  studentId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Student',
    required: true,
  },
  status: {
    type: String,
    required: true,
    enum: ['Present', 'Absent', 'Late', 'Excused'],
    default: 'Present',
  },
  note: {
    type: String,
    trim: true,
  },
});

const DailyLogSchema = new mongoose.Schema({
  day: {
    type: Number,
    required: true,
    min: 1,
    max: 31,
  },
  submittedBy: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
  },
  records: {
    type: [AttendanceRecordSchema],
    validate: [
      (arr) => arr.length > 0,
      'Daily attendance records list cannot be empty.'
    ],
  },
});

const AttendanceBucketSchema = new mongoose.Schema(
  {
    madrasaId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Madrasa',
      required: true,
    },
    classId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Class',
      required: true,
    },
    monthYear: {
      type: String,
      required: true,
      validate: {
        validator: (v) => /^\d{4}-\d{2}$/.test(v),
        message: (props) => `${props.value} is not a valid YYYY-MM month format.`
      },
    },
    dailyLogs: {
      type: [DailyLogSchema],
      default: [],
    },
  },
  {
    timestamps: true,
  }
);

// Compound Unique Index: Enforce exactly one bucket document per class per month
AttendanceBucketSchema.index({ madrasaId: 1, classId: 1, monthYear: 1 }, { unique: true });

// Compound Index: Optimize student lookup queries inside class buckets
AttendanceBucketSchema.index({ 'dailyLogs.records.studentId': 1, monthYear: 1 });

module.exports = mongoose.model('Attendance', AttendanceBucketSchema);

File: frontend/lib/models/attendance.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'attendance.freezed.dart';
part 'attendance.g.dart';

@freezed
class AttendanceRecord with _$AttendanceRecord {
  const factory AttendanceRecord({
    String? studentId,
    @Default('Present') String status,
    String? note,
  }) = _AttendanceRecord;

  factory AttendanceRecord.fromJson(Map<String, dynamic>? json) =>
      _$AttendanceRecordFromJson(json ?? {});
}

@freezed
class DailyLog with _$DailyLog {
  const factory DailyLog({
    required int day,
    String? submittedBy,
    required List<AttendanceRecord> records,
  }) = _DailyLog;

  factory DailyLog.fromJson(Map<String, dynamic>? json) =>
      _$DailyLogFromJson(json ?? {});
}

@freezed
class AttendanceBucket with _$AttendanceBucket {
  const factory AttendanceBucket({
    required String madrasaId,
    required String classId,
    required String monthYear,
    @Default([]) List<DailyLog> dailyLogs,
  }) = _AttendanceBucket;

  factory AttendanceBucket.fromJson(Map<String, dynamic>? json) =>
      _$AttendanceBucketFromJson(json ?? {});
}

@freezed
class AttendanceSummary with _$AttendanceSummary {
  const factory AttendanceSummary({
    @Default(0) int totalDaysLogged,
    @Default(0) double attendanceRate,
    @Default(Amounts()) Amounts counts,
  }) = _AttendanceSummary;

  factory AttendanceSummary.fromJson(Map<String, dynamic>? json) =>
      _$AttendanceSummaryFromJson(json ?? {});
}

@freezed
class Amounts with _$Amounts {
  const factory Amounts({
    @Default(0) @JsonKey(name: 'Present') int present,
    @Default(0) @JsonKey(name: 'Absent') int absent,
    @Default(0) @JsonKey(name: 'Late') int late,
    @Default(0) @JsonKey(name: 'Excused') int excused,
  }) = _Amounts;

  factory Amounts.fromJson(Map<String, dynamic>? json) =>
      _$AmountsFromJson(json ?? {});
}
